src/examples/AppUartModule.cpp

```diff
@@ -0,0 +1,270 @@
+                    SIZEOF_APP_UART_MODULE_TERMINAL_RESPONSE_MESSAGE, false);
+
+                if (data->splitHeader != MessageType::SPLIT_WRITE_CMD_END)
+                    return;
+#if IS_ACTIVE(LOGGING)
+                char temp[data->partLen + 1];
+                CheckedMemcpy(temp, data->data, data->partLen);
+                temp[data->partLen] = '\0';
+                logt("APPUART", "Got command one message with %s", temp);
+#endif
+                lineToReadAvailable = true;
+                GS->terminal.CheckAndProcessLine();
+                lineToReadAvailable = false;
+                readBufferOffset = 0;
+            }
+        }
+    }
+}
+
+NodeId AppUartModule::GetMeshAccessConnectionInNodeId(bool isVirtual) const
+{
+    MeshAccessConnections connIn = GS->cm.GetMeshAccessConnections(ConnectionDirection::DIRECTION_IN);
+    if (connIn.count < 1)
+        return NODE_ID_INVALID;
+    return isVirtual ? connIn.handles[0].GetVirtualPartnerId() : connIn.handles[0].GetPartnerId();
+}
+
+bool AppUartModule::IsConnectSmartPhone()
+{
+    const NodeId partnerId = GetMeshAccessConnectionInNodeId();
+    return partnerId >= NODE_ID_APP_BASE && partnerId < NODE_ID_GLOBAL_DEVICE_BASE;
+}
+
+void AppUartModule::SendAppLogQueue()
+{
+    if (logQueue._numElements < 2 || !IsConnectSmartPhone())
+        return;
+
+    AppUartLogRemain* logRemain = reinterpret_cast<AppUartLogRemain*>(logQueue.PeekNext().data);
+    char* log = reinterpret_cast<char*>(logQueue.PeekNext(1).data);
+    const u16 logLen = logRemain->logLen;
+    AppUartModuleTerminalLogMessage message;
+    if (logLen <= DATA_MAX_LEN) {
+        message.splitHeader = MessageType::SPLIT_WRITE_CMD_END;
+        message.splitCount = 0;
+        message.partLen = logLen;
+        CheckedMemcpy(message.data, log, logLen);
+    }
+    const u16 remainLogLen = logLen - logRemain->sentLength;
+    if (logLen > DATA_MAX_LEN) {
+        message.splitHeader
+            = remainLogLen > DATA_MAX_LEN ? MessageType::SPLIT_WRITE_CMD : MessageType::SPLIT_WRITE_CMD_END;
+        message.splitCount = (u8)(logRemain->sentLength / DATA_MAX_LEN);
+        message.partLen = remainLogLen > DATA_MAX_LEN ? DATA_MAX_LEN : remainLogLen;
+        CheckedMemcpy(message.data, &log[logRemain->sentLength], message.partLen);
+    }
+
+    const NodeId virtualPartnerId = GetMeshAccessConnectionInNodeId(true);
+    SendModuleActionMessage(MessageType::MODULE_ACTION_RESPONSE, virtualPartnerId,
+        AppUartModuleActionResponseMessages::RECEIVE_LOG, 0, reinterpret_cast<u8*>(&message),
+        SIZEOF_APP_UART_MODULE_TERMINAL_COMMAND_MESSAGE_STATIC + message.partLen, false);
+    if (remainLogLen > DATA_MAX_LEN) {
+        logRemain->sentLength += DATA_MAX_LEN;
+        return;
+    }
+    // send log completely
+    DiscardLogQueue();
+}
+
+bool AppUartModule::PutAppLogQueue(const char* log, u16 length)
```

Comment:
There was a lack of consideration for the critical section. Thank you for pointing this out, and is it correct to enclose it with `CRITICAL_REGION_ENTER()` and `CRITICAL_REGION_EXIT()`?

Reply:
I'm not sure what needs to be fixed. I'm trying to determine if I can't queue any more packets by checking if `Put() in the PacketQueue class returns false.
Does this mean that I should use a different method to determine in advance if there are enough buffers in the queue?

Reply:
After some research, I think I should have used IRQ instead of inline assembly.
Since the `TimerEventHandler` that calls `SendAppLogQueue` has its priority set to `NRF_SDH_BLE_STACK_OBSERVER_PRIO`, I think that calling `SetPendingEventIRQ` before queuing in `PutAppLogQueue` will prevent `SendAppLogQueue` from interrupting the queueing.
However, there are other places where `SetPendingEventIRQ` is called, but nowhere does it call `NvicClearPendingIRQ`.
If there are any mistakes in this response, please point them out to me.

```c
bool AppUartModule::PutAppLogQueue(const char* log, u16 length)
{
...

    FruityHal::SetPendingEventIRQ();
    AppUartLogRemain remain = { .sentLength = 0, .logLen = length };
    if (!logQueue.Put(reinterpret_cast<u8*>(&remain), sizeof(AppUartLogRemain)))
        return false;
    if (!logQueue.Put(reinterpret_cast<u8*>(const_cast<char*>(log)), length)) {
        logQueue.DiscardLast();
        return false;
    }
...
}
```

Reply:
Implementing critical sections is not easy to get right, the `CRITICAL_REGION_ENTER` and `..._EXIT` macros effectively forward to functionality implemented in the SoftDevice for nested critical section, otherwise only non-nested sections are supported.

Having thought about this some more, I don't think that the critical section is required here, as long as nothing is queued while a message is in the process of being queued on the connection or queued messages exist on the connection.

I would add two private members to the AppUartModule ...

```c++
u32 appUartUniqueConnectionId = 0xFFFFFFFF;
std::atomic<bool> sendingDataFlag = {};
```

... where `appUartUniqueConnectionId` will contain the unique connection id of the MeshAccessConnection, which can be retrieved in the `MeshMessageReceivedHandler` when the AppUart-specific handshake message is received. The `sendingDataFlag` will indicate if the `AppUartModule` is currently calling `SendModuleActionMessage`.

Then we wrap `SendModuleActionMessage` into a private method ...

```c++
ErrorTypeUnchecked AppUartModule::DoSendModuleActionMessage(MessageType messageType, NodeId toNode, u8 actionType, u8 requestHandle, const u8* additionalData, u16 additionalDataSize, bool reliable)
{
    sendingDataFlag.store(true);
    ErrorTypeUnchecked error = SendModuleActionMessage(messageType, toNode, actionType, requestHandle, additionalData, additionalDataSize, reliable);
    sendingDataFlag.store(false);
    return error;
}
```

... which turns the flag on before any message is enqueued and off after it was queued.

Finally we check in `PutAppLogQueue` that it is *a)* not interrupting the message-queuing (e.g. when `logt` is called with an enabled log-tag in `SendModuleActionMessage`) and *b)* that there are no messages waiting in the queue(s) of the connection to the smartphone.

```c++
bool AppUartModule::PutAppLogQueue(const char* log, u16 length)
{
    if (log == NULL || length == 0 || !IsConnectSmartPhone())
        return false;

    if (sendingDataFlag.load())
        return false;

    auto connectionHandle = ConnectionManager::GetInstance().GetConnectionByUniqueId(appUartUniqueConnectionId);
    if (connectionHandle && connectionHandle.GetConnection()->queue.GetAmountOfPackets() > 0)
        return false;

    AppUartLogRemain remain = { .sentLength = 0, .logLen = length };
    if (!logQueue.Put(reinterpret_cast<u8*>(&remain), sizeof(AppUartLogRemain)))
        return false;
    if (!logQueue.Put(reinterpret_cast<u8*>(const_cast<char*>(log)), length)) {
        logQueue.DiscardLast();
        return false;
    }
    return true;
}
```

Does this make sense?

Reply:
Thank you for the detailed explanation.I understood that I didn't need to use the critical section if I applied what you pointed out [here](https://github.com/mwaylabs/fruitymesh/pull/180#discussion_r687642118). I'm sorry for my lack of understanding.

I think I have already implemented what you explained above in [commit b66985e](https://github.com/mwaylabs/fruitymesh/pull/180/commits/b66985e542fbba9e67f5b6586170f6d9694654c9).
However, my implementation stores the `connectionHandle` of the `MeshAccessConnectionHandle` instead of the `UniqueConnectionId`.Also, multiple handles are held so that connections can be made from multiple smartphones.
If there is a problems with the way the `connectionHandle` is used, I will modify it to save the `UniqueConnectionId`. 

The method of wrapping SendModuleActionMessage in a private method is very straightforward and I will use it. (I think it was a bit confusing in my implementation)

I think I've fixed all the issues you pointed out. I think it works as expected when I actually tried to run it.If anything is missing, please let me know